bigquery_etl/backfill/shredder_mitigation.py

```diff
@@ -0,0 +1,207 @@
+        return DataTypeGroup.STRING
+    elif isinstance(value, NoneType):
+        return DataTypeGroup.UNDETERMINED
+    else:
+        raise ValueError(f"Unsupported data type: {type(value)}")
+
+
+def classify_columns(
+    new_row: dict, existing_dimension_columns: list, new_dimension_columns: list
+) -> tuple[list[Column], list[Column], list[Column], list[Column], list[Column]]:
+    """Compare the new row with the existing columns and return the list of common, added and removed columns by type."""
+    common_dimensions = []
+    added_dimensions = []
+    removed_dimensions = []
+    metrics = []
+    undefined = []
+
+    if not new_row or not existing_dimension_columns or not new_dimension_columns:
+        raise click.ClickException(
+            f"Missing required parameters. Received: new_row= {new_row}\nexisting_dimension_columns= {existing_dimension_columns},\nnew_dimension_columns= {new_dimension_columns}.")
+
+    missing_dimensions = [dimension for dimension in new_dimension_columns if dimension not in new_row]
+    if not len(missing_dimensions) == 0:
+        raise click.ClickException(
+            f"Inconsistent parameters. Columns in new dimensions not found in new row: {missing_dimensions}")
+
+    for key in existing_dimension_columns:
+        if key not in new_dimension_columns:
+            removed_dimensions.append(
+                Column(
+                    key,
+                    DataTypeGroup.UNDETERMINED,
+                    ColumnType.UNDETERMINED,
+                    ColumnStatus.REMOVED,
+                )
+            )
+
+    for key, value in new_row.items():
+        value_type = get_bigquery_type(value)
+        if key in existing_dimension_columns:
+            common_dimensions.append(
+                Column(
+                    key,
+                    value_type,
+                    ColumnType.DIMENSION,
+                    ColumnStatus.COMMON,
+                )
+            )
+        elif key not in existing_dimension_columns and key in new_dimension_columns:
+            added_dimensions.append(
+                Column(
+                    key,
+                    value_type,
+                    ColumnType.DIMENSION,
+                    ColumnStatus.ADDED,
+                )
+            )
+        elif (
+            key not in existing_dimension_columns
+            and key not in new_dimension_columns
+            and (value_type is DataTypeGroup.NUMERIC or value_type is DataTypeGroup.FLOAT)
+        ):
+            # Columns that are not in the previous or new list of grouping columns are metrics.
+            metrics.append(
+                Column(
+                    key,
+                    get_bigquery_type(value),
+                    ColumnType.METRIC,
+                    ColumnStatus.COMMON,
+                )
+            )
+        else:
+            undefined.append(
+                Column(
+                    key,
+                    get_bigquery_type(None),
```

Comment:
not sure I totally get why you default to UNDEFINED here ... 
wouldn't it be helpful for debugging to see that it's a string or a numeric value?

Reply:
The default to UNDETERMINED is for metric/dimension which I couldn't be identified. The key and value would show, it's all the info available.
-- I did change that I was calling the function instead of `value_type`.